tests/test_cliCmdQueryCollections.py

```diff
@@ -56,41 +75,159 @@ def test_all(self):
+            self.assertAstropyTablesEqual(readTable(result.output), expected)
 
             # Verify that with a glob argument, that only collections whose
             # name matches with the specified pattern are returned.
-            result = runner.invoke(cli, ["query-collections", "here", "t*"])
-            self.assertEqual({"collections": [tag]}, yaml.safe_load(result.output))
+            result = self.runner.invoke(cli, ["query-collections", "here", "t*"])
+            self.assertEqual(result.exit_code, 0, clickResultMsg(result))
+            expected = Table((("tag",), ("TAGGED",)),
+                             names=("Name", "Type"))
+            self.assertAstropyTablesEqual(readTable(result.output), expected)
 
             # Verify that with a collection type argument, only collections of
             # that type are returned.
-            result = runner.invoke(cli, ["query-collections", "here", "--collection-type", "RUN"])
-            self.assertEqual({"collections": [run]}, yaml.safe_load(result.output))
+            result = self.runner.invoke(cli, ["query-collections", "here", "--collection-type", "RUN"])
+            self.assertEqual(result.exit_code, 0, clickResultMsg(result))
+            expected = Table((("ingest/run",), ("RUN",)),
+                             names=("Name", "Type"))
+            self.assertAstropyTablesEqual(readTable(result.output), expected)
+
+
+class ChainedCollectionsTest(ButlerTestHelper, unittest.TestCase):
+
+    @staticmethod
+    def _mock_export(refs: Iterable[DatasetRef], *,
+                     directory: Optional[str] = None,
+                     transfer: Optional[str] = None) -> Iterable[FileDataset]:
+        """A mock of `Datastore.export` that satisfies the requirement that
+        the refs passed in are included in the `FileDataset` objects
+        returned.
+
+        This can be used to construct a `Datastore` mock that can be used
+        in repository export via::
+
+            datastore = unittest.mock.Mock(spec=Datastore)
+            datastore.export = _mock_export
+
+        """
+        for ref in refs:
+            yield FileDataset(refs=[ref],
+                              path="mock/path",
+                              formatter="lsst.daf.butler.formatters.json.JsonFormatter")
+
+    @staticmethod
+    def _mock_get(ref: DatasetRef, parameters: Optional[Mapping[str, Any]] = None
+                  ) -> Tuple[int, Optional[Mapping[str, Any]]]:
+        """A mock of `Datastore.get` that just returns the integer dataset ID
+        value and parameters it was given.
+        """
+        return (ref.id, parameters)
+
+    def setUp(self):
+        self.runner = LogCliRunner()
+
+    def testChained(self):
+        with self.runner.isolated_filesystem():
+
+            # Create a butler and add some chained collections
+            butlerCfg = Butler.makeRepo("here")
+            with unittest.mock.patch.object(Datastore, "fromConfig", spec=Datastore.fromConfig):
```

Comment:
I find the `with` a bit confusing here since it implies that something is freed or reversed when the block exits. This seems to be how it works though.

Reply:
I didn't totally understand it (it's cargo culled from something @TallJimbo wrote), so I did more research. The deal is that `Datastore.fromConfig` gets patched: replaced with a function that returns a `magicmock`. The butler initializer calls `Datastore.fromConfig`, which is a `MagicMock` instance and calling it returns a new `MagicMock` instance, so now that butler's `self.datastore` is a magicmock. Init finishes, and `export` and `get` are monkey patched onto the `butler.datastore` magic mock. Then the `with` block exits, which does restore `Datastore.fromConfig` to its normal function, but `butler.datastore` is of course not changed because how would it and anyway we don't want it to be - it's why we did all the above anyway.

Reply:
However, all that seems kind of complicated and it seems like it's not necessary to patch `Datastore.fromConfig` - the butler init works just fine with it *not* patched. We do run into problems when we call `butler1.import_(filename=os.path.join(TESTDIR, "data", "registry", "base.yaml"))` without patching, but we can simply replace the `datastore` functions that need replacing, without any context manager. Doing this seems to be enough:

            butler1.datastore.export = self._mock_export
            butler1.datastore.get = self._mock_get
            butler1.datastore.ingest = MagicMock()

(note that instead of `datastore` be a MagicMock with `export` and `get` monkey patched on, I've replaced those two with our alternative impls, and made `ingest` a mock because it will get called in the course of calling `butler.ingest_`, but does not seem to need to do anything.

Reply:
@TallJimbo if there's a reason to use the patch context manager, please let me know? In the meantime I'll change it to the simpler version I just described and you and @timj can 👍 or 👎 that impl if y'all want.

Reply:
(hmm, it seems to need the context manager in test_simpleButler. working on understanding why...)

Reply:
aha. The butlers were getting created slightly differently, the one that worked was being init'd with the config returned by `Butler.makeRepo`, and the one that did not was building a config itself, and needed `config["root"]` to be set.

Reply:
This was the first time I ever really used `unittest.mock` (other than small changes to tests others had written), so if you find a simpler way to accomplish what my code did, assume I just wasn't aware of that alternative.